### PR TITLE
feat: implement jwt

### PR DESCRIPTION
**Pull request by Ethel3**

### CODE REVIEW

File: controllers/game.controller.js [hunk 1]
```diff
@@ -47,15 +66,22 @@ export const updateGame = async (req, res, next) => {
...
+    return res.status(403).json({ message: 'Only admins can update game' });
+  }
+}  catch (error) {
     next(error);
   }
 };
 
+/*
+//TODO: Only allow admins to delete games
+*/
```
Comment (kwakubiney) [inline controllers/game.controller.js:64]: Eliminate the `Todo` once the it has been done.

Comment (Ethel3) [reply controllers/game.controller.js:64]: Alright.

File: controllers/user.controller.js [hunk 2]
```diff
@@ -35,28 +66,41 @@ export const updateUser = async (req, res, next) => {
...
       UPDATE users
-      SET username = $1, age = $2, email = $3
-      WHERE id = $4
+      SET username = $1, age = $2, email = $3, role = $4, password = $5
+      WHERE id = $6
       RETURNING *;
     `;
-    const updateUserValues = [username, age, email, id];
+
+    const updateUserValues = [username, age, email, role, hashedPassword, id];
```
Comment (kwakubiney) [inline controllers/user.controller.js]: Why are we retrieving the password on an account update? I don't think that's necessary. Also, you do not want to send `password` in a request body, after account creation. That's what JWTs are for. Passwords aren't needed to verify the user after the user has already signed in

File: controllers/user.controller.js [hunk 3]
```diff
@@ -88,12 +132,78 @@ export const getUsers = async (req, res, next) => {
...
   } catch (error) {
     next(error);
   }
 };
 
+export const loginUser = async (req, res, next) => {
+  try {
+    const { username, password } = req.body;
+    console.log('Received username:', username);
+    console.log('Received password:', password);
```
Comment (kwakubiney) [inline controllers/user.controller.js]: These are irrelevant. Please remove

File: controllers/user.controller.js [hunk 4]
```diff
@@ -88,12 +132,78 @@ export const getUsers = async (req, res, next) => {
...
+
+
+const saltRounds = 10; 
+const hashedAdminPassword = bcrypt.hashSync(adminPassword, saltRounds);
+
+console.log('Hashed Admin Password:', hashedAdminPassword);
+    
+    if (!password) {
+      return res.status(401).json({ message: 'Invalid credentials' });
+    }
```
Comment (kwakubiney) [inline controllers/user.controller.js]: Nit: The code isn't aligned properly. Too much irregular spacing. Fix that

File: controllers/usergame.controller.js [hunk 5]
```diff
@@ -3,6 +3,11 @@ import pool from '../dbConfig.mjs';
 export const createUserGame = async (req, res, next) => {
```
Comment (kwakubiney) [inline controllers/usergame.controller.js]: I think we should rename this to `createGameForUser`

File: controllers/usergame.controller.js [hunk 6]
```diff
@@ -59,7 +69,11 @@ export const updateUserGame = async (req, res, next) => {
 export const deleteUserGame = async (req, res, next) => {
   try {
     const { id } = req.params;
-
+    if(req.id !=  user_id){
+      return res.json({
+        "message": "User is not authorized to perform this operation"
+      })
+  }
```
Comment (kwakubiney) [inline controllers/usergame.controller.js]: On subsequent tasks, we need to find a way to refactor this `if(req.id != user_id)` check. We could move it to a separate function or possibly to some middleware later

File: middleware/auth.js [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  }
+
+    const decoded = jwt.verify(token, process.env("SECRET"));
+    req.user = decoded;
+    next();
+  } catch (err) {
+    res.status(401).json({ message: 'User not authorized to access this resource' });
+  }
+};
+// check if the user is an admin
```
Comment (kwakubiney) [inline middleware/auth.js]: Remove this comment. Isn't totally helpful IMO

File: middleware/auth.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+export const isAdmin = (req, res, next)=> {
+  if (req.user && req.user.role === 'admin') {
+    return next();
+  } else {
+    return res.status(403).json({ message: 'Only admins can perform this operation' });
```
Comment (kwakubiney) [inline middleware/auth.js]: If the admin logic isn't included in the task, let's remove all occurrences of it in the codebase, please.

File: routes/game.routes.js [hunk 9]
```diff
@@ -1,11 +1,12 @@
...
+import { authMiddleware, isAdmin } from '../middleware/auth.js';
 
 const routes = express.Router();
 
-routes.post('/', createGame);
-routes.put('/:id', updateGame);
-routes.delete('/:id', deleteGame);
+routes.post('/',authMiddleware, isAdmin, createGame);
+routes.put('/:id',authMiddleware, isAdmin, updateGame);
+routes.delete('/:id',authMiddleware, isAdmin, deleteGame);
```
Comment (kwakubiney) [inline routes/game.routes.js]: `isAdmin` functionality is not included in this task. Please remove.

**Pull request merged**